src/hdmf/data_utils.py

```diff
@@ -286,69 +314,140 @@ def __init__(self, **kwargs):
+    def estimate_chunk_shape(**kwargs) -> Tuple[int, ...]:
+        """Select chunk shape with size in MB less than the threshold of chunk_mb."""
+        maxshape, itemsize, chunk_mb, chunk_shape_constraints, chunking_strategy = getargs(
+            "maxshape", "itemsize", "chunk_mb", "chunk_shape_constraints", "chunking_strategy", kwargs
+        )
         assert chunk_mb > 0, f"chunk_mb ({chunk_mb}) must be greater than zero!"
+        assert chunking_strategy in ['axis_ratio', 'uniform'], (
+            "Unrecognized `chunking_strategy` selected! Please select either 'axis_ratio' or 'uniform'."
+        )
+
+        maxshape = np.array(maxshape)
+        number_of_dimensions = len(maxshape)
+        if chunk_shape_constraints is not None:
+            assert len(chunk_shape_constraints) == len(maxshape)
+            chunk_shape_constraints = np.array(chunk_shape_constraints)
+        else:
+            chunk_shape_constraints = np.array([None for _ in range(number_of_dimensions)])
 
-        n_dims = len(self.maxshape)
-        itemsize = self.dtype.itemsize
         chunk_bytes = chunk_mb * 1e6
 
-        min_maxshape = min(self.maxshape)
-        v = tuple(math.floor(maxshape_axis / min_maxshape) for maxshape_axis in self.maxshape)
-        prod_v = math.prod(v)
-        while prod_v * itemsize > chunk_bytes and prod_v != 1:
-            non_unit_min_v = min(x for x in v if x != 1)
-            v = tuple(math.floor(x / non_unit_min_v) if x != 1 else x for x in v)
+        none_axes = np.where([x is None for x in chunk_shape_constraints])[0]
+        constrained_axes = np.where([x is not None for x in chunk_shape_constraints])[0]
+        if chunking_strategy == "uniform":
+            estimated_chunk_shape = np.array(chunk_shape_constraints)
+            capped_axes = none_axes
+            while any(capped_axes):
+                number_of_free_axes = len(none_axes)
+
+                # Estimate the amount to fill uniformly across the unconstrained axes
+                # Note that math.prod is 1 if all axes are None
+                constrained_bytes = math.prod(estimated_chunk_shape[constrained_axes]) * itemsize
+                estimated_fill_factor = math.floor((chunk_bytes / constrained_bytes) ** (1 / number_of_free_axes))
+
+                # Update axes and constraints in the event that the fill factor pushed some axes beyond their maximum
+                capped_axes = none_axes[np.where(maxshape[none_axes] <= estimated_fill_factor)[0]]
+                estimated_chunk_shape[capped_axes] = maxshape[capped_axes]  # Cap the estimate at the max
+                chunk_shape_constraints[capped_axes] = maxshape[capped_axes]  # Consider capped axis a constraint
+                none_axes = np.where([x is None for x in chunk_shape_constraints])[0]
+                constrained_axes = np.where([x is not None for x in chunk_shape_constraints])[0]
+            estimated_chunk_shape[none_axes] = estimated_fill_factor
```

Comment:
OK, I think this seems like a good alternative solution, which should achieve the same answer in far fewer loops, but I agree with @oruebel that it could go in a separate function. I liked the way I broke it apart where the math was in a separate function that was just trying to achieve a product of a vector (not knowing that it was related to the size and shape of an array).

Also, a minor point: I would remove the `np.where`s. If you are going this way you can just index with the boolean np arrays, which will make the syntax a bit cleaner.